lock-api/src/main/java/com/palantir/lock/client/AsyncTimeLockUnlocker.java

```diff
@@ -0,0 +1,114 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+import java.time.Duration;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.logsafe.SafeArg;
+
+/**
+ * Releases lock tokens from a {@link TimelockService} asynchronously.
+ *
+ * There is another layer of retrying below us (at the HTTP client level) for external timelock users.
+ * Also, in the event we fail to unlock (e.g. because of a connection issue), locks will eventually time-out.
+ * Thus not retrying is reasonably safe (as long as we can guarantee that the lock won't otherwise be refreshed).
+ */
+public class AsyncTimeLockUnlocker {
+    private static final Logger log = LoggerFactory.getLogger(AsyncTimeLockUnlocker.class);
+
+    private static final Duration KICK_JOB_INTERVAL = Duration.ofSeconds(1);
+
+    private final TimelockService timelockService;
+    private final ScheduledExecutorService scheduledExecutorService;
+
+    private final AtomicBoolean available = new AtomicBoolean(true);
+    private final AtomicReference<Set<LockToken>> outstandingLockTokens;
+
+    AsyncTimeLockUnlocker(TimelockService timelockService, ScheduledExecutorService scheduledExecutorService) {
+        this(timelockService, scheduledExecutorService, new AtomicReference<>(ImmutableSet.of()));
+    }
+
+    @VisibleForTesting
+    AsyncTimeLockUnlocker(TimelockService timelockService,
+            ScheduledExecutorService scheduledExecutorService,
+            AtomicReference<Set<LockToken>> outstandingLockTokens) {
+        this.timelockService = timelockService;
+        this.scheduledExecutorService = scheduledExecutorService;
+        this.outstandingLockTokens = outstandingLockTokens;
+        schedulePeriodicKickJob();
+    }
+
+    /**
+     * Adds all provided lock tokens to a queue to eventually be scheduled for unlocking.
+     * Locks in the queue are unlocked asynchronously, and users must not depend on these locks being unlocked /
+     * available for other users immediately.
+     *
+     * @param tokens Lock tokens to schedule an unlock for.
+     */
+    public void enqueue(Set<LockToken> tokens) {
+        outstandingLockTokens.getAndAccumulate(tokens, Sets::union);
```

Comment:
What's the rationale for this vs. just using a concurrent set implementation and adding the elements directly? 

I'm worried about the number of set iterators we might have sitting with this implementation - could be dozens of unioned and diffed Sets in here. Of course, we'd have to be more careful about read/write contention in this class with the other approach.

Reply:
Fair enough - main thing was that this gives a very easy way of draining the set in `unlockOutstanding`. Discussed offline - we now have a solution that uses read-write locks to maintain high concurrency in adding to a straight concurrentset, and keeps the synchronized bits small.